util/dvsim/Deploy.py

```diff
@@ -279,118 +277,138 @@ def dispatch_cmd(self):
-            grep_cmd = "grep -m 1 -A 4 -E \'" + fail_pattern + "\' " + self.log
-            (status, rslt) = subprocess.getstatusoutput(grep_cmd)
-            if rslt:
-                msg = "```\n{}\n```\n".format(rslt)
-                self.fail_msg += msg
-                log.log(VERBOSE, msg)
-                seen_fail_pattern = True
-                break
-
-        if seen_fail_pattern:
+        # Only one pattern need to be seen.
+        failed = False
+        chk_failed = bool(self.fail_patterns)
+
+        # All pass patterns need to be seen - hence maintain it as bool dict.
+        passed = {p: False for p in self.pass_patterns}
+        chk_passed = bool(self.pass_patterns)
+
+        try:
+            f = open(self.log, "r", encoding="UTF-8")
+        except Exception as e:
+            msg = "Error opening file \"{!r}\":\n{!r}".format(self.log, e)
+            self.fail_msg += msg
+            log.log(VERBOSE, msg)
+            return False
+
+        lines = f.readlines()
+
+        def _find_patterns(patterns, line):
+            '''Helper function that returns true if all or any of the given
+            patterns is found, else False.'''
+
+            assert patterns
+            for pattern in patterns:
+                match = re.search(r"{}".format(pattern), line)
+                if match:
+                    return pattern
+            return None
+
+        def _get_n_lines(pos, num):
+            "Helper function that returns next N lines starting at pos index."
+
+            idx = min(len(lines) - 1, pos + num - 1)
+            return ''.join(lines[pos:idx]).strip()
+
+        if chk_failed or chk_passed:
+            for cnt, line in enumerate(lines):
+                if chk_failed:
+                    if _find_patterns(self.fail_patterns, line) is not None:
+                        # Print 4 additional lines to help debug more easily.
+                        msg = "```\n{}\n```\n".format(_get_n_lines(cnt, 5))
+                        self.fail_msg += msg
+                        log.log(VERBOSE, msg)
+                        failed = True
+                        chk_failed = False
+
+                if chk_passed:
+                    pattern = _find_patterns(self.pass_patterns, line)
+                    if pattern is not None:
+                        passed[pattern] = True
+                        chk_passed = not all(passed.values())
+
+        # If failed, then nothing else to do. Just return.
+        if failed:
             return False
 
         # If no fail patterns were seen, but the job returned with non-zero
         # exit code for whatever reason, then show the last 10 lines of the log
         # as the failure message, which might help with the debug.
         if self.process.returncode != 0:
-            msg = "Last 10 lines of the log:<br>\n"
+            msg = "Last 10 lines:\n```\n{}\n```\n".format(''.join(
+                lines[-10:]).strip())
```

Comment:
This sort of thing is why I don't like auto-formatters. It looks horrible! If we must use the auto-formatter, could you put the argument to `format` into a local variable so that the lines break in a slightly more sensible place?

The same comment goes for line 390 below.

Reply:
Looks good, thanks.